packages/frontend/electron/src/main/windows-manager/main-window.ts

```diff
@@ -227,27 +227,50 @@ export async function showMainWindow() {
   window.focus();
 }
 
+let hiddenWindowPromise: Promise<BrowserWindow> | undefined;
+
+const getWindowAdditionalArguments = async () => {
+  const { getExposedMeta } = await import('../exposed');
+  const mainExposedMeta = getExposedMeta();
+  return [
+    `--main-exposed-meta=` + JSON.stringify(mainExposedMeta),
+    `--window-name=hidden-window`,
+  ];
+};
+
 /**
  * Open a URL in a hidden window.
  */
 export async function openUrlInHiddenWindow(url: string) {
-  const win = new BrowserWindow({
-    width: 1200,
-    height: 600,
-    webPreferences: {
-      preload: join(__dirname, './preload.js'),
-    },
-    show: environment.isDebug,
-  });
-  win.on('close', e => {
-    e.preventDefault();
-    if (!win.isDestroyed()) {
-      win.destroy();
-    }
-  });
+  if (!hiddenWindowPromise) {
+    const { promise, resolve, reject } = Promise.withResolvers<BrowserWindow>();
+
+    hiddenWindowPromise = promise;
+
+    const win = new BrowserWindow({
+      width: 1200,
+      height: 600,
+      webPreferences: {
+        preload: join(__dirname, './preload.js'),
+        additionalArguments: await getWindowAdditionalArguments(),
+      },
+      show: false,
+    });
+    win.on('close', e => {
+      e.preventDefault();
+      if (win && !win.isDestroyed()) {
+        win.destroy();
+      }
+    });
+    win.on('unresponsive', reject);
+    win.on('closed', () => {
+      hiddenWindowPromise = undefined;
+    });
+    resolve(win);
+  }
+  const win = await hiddenWindowPromise;
```

Comment:
I'm curious when will the hidden be closed?

Reply:
it shhould be closed when the main window is closed. https://github.com/toeverything/AFFiNE/blob/f544e69d0234afa3c49fedb2f38ec64149594382/packages/frontend/electron/src/main/windows-manager/main-window.ts#L118

Reply:
I think in the current workflow the hidden window will only stay in the magic link route.
A possible solution (which was implemented before but was removed) is to signal the main window when login process in the hidden window is done and then close the hidden window.

Reply:
it's better to have

Reply:
ok, I will add back mentioned logic